### PR TITLE
feat: add Action overloads to ConfigureFunctionsWebApplication for FunctionsApplicationBuilder

### LINKED ISSUE #3052: Add ConfigureFunctionsWebApplication extension allowing to pass configure action
### Description

Currently F# ignores Extension Startups that are generated ([issue #1246](https://github.com/Azure/azure-functions-dotnet-worker/issues/1246)) and as a workaround it must be configured manually.

It is easily doable for classic `IHostBuilder`, because its extension allows to pass configure action:
https://github.com/Azure/azure-functions-dotnet-worker/blob/cfb4d9cba288df1bf1e14572ac5af46600cfd6d3/extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsHostBuilderExtensions.cs#L37

But this is not possible for the new preferred `IHostApplicationBuilder` as the only extension method does not allow to pass configure action:
https://github.com/Azure/azure-functions-dotnet-worker/blob/cfb4d9cba288df1bf1e14572ac5af46600cfd6d3/extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs#L18C5-L18C121

It makes .NET Aspire harder to use, because its entire `ServiceDefaults` rely on `IHostApplicationBuilder` and it requires either custom implementation or below hack to access internal `HostBuilder` property:
```
    let builder = FunctionsApplication.CreateBuilder(args)

    builder
        .GetType()
        .GetProperty(
            "HostBuilder",
            System.Reflection.BindingFlags.NonPublic
            ||| System.Reflection.BindingFlags.Instance
        )
        .GetValue(builder, null)
    :?> IHostBuilder
    |> _.ConfigureFunctionsWebApplication(ServiceBusExtensionStartup().Configure)
    |> ignore

    builder.AddServiceDefaults().Build().Run()
```

### PR DESCRIPTION
**Pull request by MO2k4**
## Summary

- Adds `Action<IFunctionsWorkerApplicationBuilder>` and `Action<HostBuilderContext, IFunctionsWorkerApplicationBuilder>` overloads to `FunctionsApplicationBuilderAspNetCoreExtensions`, mirroring the existing `IHostBuilder` API in `FunctionsHostBuilderExtensions`
- Both overloads delegate through `FunctionsApplicationBuilder.HostBuilder` — zero duplicated logic; idempotency and ASP.NET Core integration guards continue to run through the single existing implementation
- Includes null guards (`ArgumentNullException.ThrowIfNull`) on the callback parameters

Fixes #3052

## Test Plan

- [x] 3 new tests added: action callback, context+action callback, idempotency (service descriptor count + callback invocation count)
- [x] All 26 tests pass (`dotnet test test/extensions/Worker.Extensions.Http.AspNetCore.Tests/`)
- [x] Build clean: 0 errors, 0 warnings on net6.0 and net8.0

### CODE REVIEW

File: extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs [hunk 1]
```diff
@@ -20,4 +21,30 @@ public static FunctionsApplicationBuilder ConfigureFunctionsWebApplication(this
...
+
+    /// <summary>
+    /// Configures the worker to use the ASP.NET Core integration, enabling advanced HTTP features.
+    /// </summary>
+    /// <param name="builder">The <see cref="FunctionsApplicationBuilder"/> to configure.</param>
+    /// <param name="configureWorker">The worker configure callback.</param>
+    /// <returns>The <see cref="FunctionsApplicationBuilder"/> for chaining.</returns>
+    public static FunctionsApplicationBuilder ConfigureFunctionsWebApplication(this FunctionsApplicationBuilder builder, Action<IFunctionsWorkerApplicationBuilder> configureWorker)
+    {
+        ArgumentNullException.ThrowIfNull(configureWorker);
```
Comment (kshyju) [inline extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs:34]: Could we update the method summary to document what exception is being thrown? (Use `<exception cref="ArgumentNullException">`)

Comment (MO2k4) [reply extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs:34]: Thanks for the hint, i've added it

**Comment by MO2k4** [comment]
@kshyju I've updated the release notes as well, if you want to take a look

File: release_notes.md [hunk 2]
```diff
@@ -17,4 +17,8 @@
 ### Microsoft.Azure.Functions.Worker.Grpc <version>
 
 - Update protobuf version to v1.12.0-protofile and add support for propagating tags from the worker to the functions host (#3303).
-- Update protobuf version to v1.13.0-protofile to add support for propagating OpenTelemetry baggage to the worker (#3319).
\ No newline at end of file
+- Update protobuf version to v1.13.0-protofile to add support for propagating OpenTelemetry baggage to the worker (#3319).
+
+### Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore <version>
+
+- Add `Action<IApplicationBuilder>` overloads to `ConfigureFunctionsWebApplication` for `FunctionsApplicationBuilder` (#3052)
```
Comment (kshyju) [inline release_notes.md]: This release notes file is for the worker packages. Since you are making changes to the aspnetcore extension package, the changelog entry should go to https://github.com/Azure/azure-functions-dotnet-worker/blob/main/extensions/Worker.Extensions.Http.AspNetCore/release_notes.md file.

**Comment by kshyju** [comment]
> @kshyju I've updated the release notes as well, if you want to take a look

Thanks! Can you also review the feedback from GHCP review and address them?

File: release_notes.md (hunk 2 above)
Comment (MO2k4) [reply release_notes.md]: Ah sorry i have not seen this

File: test/extensions/Worker.Extensions.Http.AspNetCore.Tests/FunctionsApplicationBuilderAspNetCoreExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            var httpProxyingMiddlewares = builder.Services.Where(d => d.ServiceType == typeof(FunctionsHttpProxyingMiddleware)).ToList();
+            var coordinators = builder.Services.Where(d => d.ServiceType == typeof(IHttpCoordinator)).ToList();
+            Assert.Single(endpointDataSources);
+            Assert.Single(httpProxyingMiddlewares);
+            Assert.Single(coordinators);
+
+            // Also verify end-to-end resolution works
+            var host = builder.Build();
+            VerifyRegistrationOfAspNetCoreIntegrationServices(host);
+            Assert.Equal(2, callbackCount);
```
[the remark this answers is not in the document]
Comment (MO2k4) [reply test/extensions/Worker.Extensions.Http.AspNetCore.Tests/FunctionsApplicationBuilderAspNetCoreExtensionsTests.cs]: Checked and adapted

File: test/extensions/Worker.Extensions.Http.AspNetCore.Tests/FunctionsApplicationBuilderAspNetCoreExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            var httpCoordinator = host.Services.GetService<IHttpCoordinator>();
+            Assert.NotNull(httpCoordinator);
+            Assert.IsType<DefaultHttpCoordinator>(httpCoordinator);
+        }
+
+        private class TestMiddleware : IFunctionsWorkerMiddleware
+        {
+            public Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+            {
+                return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (MO2k4) [reply test/extensions/Worker.Extensions.Http.AspNetCore.Tests/FunctionsApplicationBuilderAspNetCoreExtensionsTests.cs]: Checked and adapted

File: test/extensions/Worker.Extensions.Http.AspNetCore.Tests/FunctionsApplicationBuilderAspNetCoreExtensionsTests.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            // Also verify end-to-end resolution works
+            var host = builder.Build();
+            VerifyRegistrationOfAspNetCoreIntegrationServices(host);
+            Assert.Equal(2, callbackCount);
+        }
+
+        private static void VerifyRegistrationOfCustomMiddleware(IHost host)
+        {
+            Assert.NotNull(host.Services.GetService<TestMiddleware>());
+        }
```
[the remark this answers is not in the document]
Comment (MO2k4) [reply test/extensions/Worker.Extensions.Http.AspNetCore.Tests/FunctionsApplicationBuilderAspNetCoreExtensionsTests.cs:83]: Since the middleware is directly registered before this assertion this looks like a false positive to me. 

> underTest.ConfigureFunctionsWebApplication(builder => builder.UseMiddleware<TestMiddleware>());

**Comment by MO2k4** [comment]
@kshyju I've checked the three comments by ghcp and adapted two, the first looks like a false positive to me, because it is handled directly before the assertion.

File: extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs (hunk 1 above)
Comment (kshyju) [reply extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs:34]: Could we change it to `<exception cref="ArgumentNullException">Thrown when <paramref name="configureWorker"/> is null.</exception>` ?

File: release_notes.md (hunk 2 above)
Comment (kshyju) [reply release_notes.md]: It's `IFunctionsWorkerApplicationBuilder`, not `IApplicationBuilder`

Comment (MO2k4) [reply release_notes.md]: It is adjusted and finally it is also in the right place

File: extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs (hunk 1 above)
Comment (MO2k4) [reply extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs:34]: It is adjusted

File: release_notes.md [hunk 6]
```diff
@@ -17,4 +17,5 @@
 ### Microsoft.Azure.Functions.Worker.Grpc <version>
 
 - Update protobuf version to v1.12.0-protofile and add support for propagating tags from the worker to the functions host (#3303).
-- Update protobuf version to v1.13.0-protofile to add support for propagating OpenTelemetry baggage to the worker (#3319).
\ No newline at end of file
+- Update protobuf version to v1.13.0-protofile to add support for propagating OpenTelemetry baggage to the worker (#3319).
+
```
Comment (kshyju) [inline release_notes.md]: Could you please undo the change in this file? Also can we add a test to validate the Argument null exception for the new overload? We should be good to merge this change after that.

Comment (MO2k4) [reply release_notes.md]: Oh sorry i missed this earlier, this change is now reverted and i've added the two test cases to verify the null exceptions. Should i squash/cleanup the commits?

Comment (kshyju) [reply release_notes.md]: It will be squash merged when we merge the PR. Thank you!

File: extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs [hunk 7]
```diff
@@ -20,4 +21,32 @@ public static FunctionsApplicationBuilder ConfigureFunctionsWebApplication(this
...
+    }
+
+    /// <summary>
+    /// Configures the worker to use the ASP.NET Core integration, enabling advanced HTTP features.
+    /// </summary>
+    /// <param name="builder">The <see cref="FunctionsApplicationBuilder"/> to configure.</param>
+    /// <param name="configureWorker">The worker configure callback receiving the host builder context.</param>
+    /// <returns>The <see cref="FunctionsApplicationBuilder"/> for chaining.</returns>
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="configureWorker"/> is null.</exception>
+    public static FunctionsApplicationBuilder ConfigureFunctionsWebApplication(this FunctionsApplicationBuilder builder, Action<HostBuilderContext, IFunctionsWorkerApplicationBuilder> configureWorker)
```
Comment (jviau) [inline extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsApplicationBuilderAspNetCoreExtensions.cs:46]: I recall us intentionally not adding `HostBuilderContext` to this public API for the time being. Unsure if we want to add it at this time.

**Comment by MO2k4** [comment]
@jviau yes you are right, i tried to create an example repo to verify the original issue and could not verify the claim 

https://github.com/MO2k4/verify-issue-3052

sorry for jumping directly to resolving this, without first checking if it is a real issue :-/

**Comment by jviau** [comment]
@MO2k4 no worries! Happy to see the contribution effort.

Is there anything else that is relevant from that issue or this PR? Or is it safe to close the PR?

**Comment by MO2k4** [comment]
@jviau i assume it is save to close then
